### PR TITLE
Layer shutdown support

### PR DESCRIPTION
**Pull request by afpro**
shutdown Layer, let `BackgroundTask` exit gracefully.

### CODE REVIEW

**Comment by hrxi** [comment]
First of all, thanks for implementing this. It's definitely a wanted feature.

This PR unfortunately adds a lock into the common path of writing to the log. I'm not sure if it can be done without though.

The "avoid fixed crate version" is unnecessary, a `snap = "1.0.5"` requirement is the same as a `snap = "^1.0.5"` in `Cargo.toml`.

**Comment by afpro** [comment]
> 

oh, idk crate version '^' is unnecessary, got it.
and the lock, we need find a way let `BackgroundTaks` quit, release `Sender` is the most straight forward way. i thought about the performance too, `Layer` is definitely high frequency used.
in my case, my project add a `EnvFilter` to loki layer, so this is not problem for me, but not ok for everyone. perhaps we can let shutdown `BackgroundTask` by a `Notify` or `Semaphore`?

**Comment by tiagolobocastro** [comment]
Would closing the channel help achieve the same?
Some channels (eg: async_channel) allows us to close the channel through an immutable reference to the sender, if this one doesn't make we can close the receiver side instead?

**Comment by afpro** [comment]
`async_channel` use `concurrent_queue`， `concurrent_queue` implement immutable operation with 'atomic', seems good to me. @hrxi what do you think?

**Comment by afpro** [comment]
btw, we can use `async_channel::unbounded()` for channel full problem.

**Comment by tiagolobocastro** [comment]
Depending on how the receive-side is implemented here, I probably wouldn't recommend unbounded as you could end up infinitely increasing memory usage..

**Comment by afpro** [comment]
ok, then i will update this PR with a `tokio::sync::Notify` based solution, just quit `BackgroundTask` after drain all available data in `channel`.

**Comment by afpro** [comment]
i've tested this on my own project. just manually tested, no unit test.

**Comment by hrxi** [comment]
Looks good to me. Could you add the usage of the shutdown handle to an example? I think it'd be ready to get merged then.

File: src/lib.rs [hunk 1]
```diff
@@ -632,3 +660,20 @@ impl Buffer {
         &self.snappy[..snappy_len]
     }
 }
+
+/// handler with weak reference to layer quit notify
```
Comment (tiagolobocastro) [inline src/lib.rs:664]: Should these doc comments start with `C`apital and end with `.`? will match existing doc comments on this file AFAICT

Comment (afpro) [reply src/lib.rs:664]: English is not my main language, this is my best, feel free to change after merge if code is ok.

**Comment by afpro** [comment]
> Looks good to me. Could you add the usage of the shutdown handle to an example? I think it'd be ready to get merged then.

i'll add an example today or tomorrow

**Comment by hrxi** [comment]
Thanks for your persistence. :)

**Pull request merged**